Source/stores.cpp

```diff
@@ -307,191 +315,200 @@ void StoreAutoPlace()
-	AddSText(0, 10, _("Talk to Griswold"), UiFlags::ColorBlue | UiFlags::AlignCenter, true);
-	AddSText(0, 12, _("Buy basic items"), UiFlags::ColorWhite | UiFlags::AlignCenter, true);
-	AddSText(0, 14, _("Buy premium items"), UiFlags::ColorWhite | UiFlags::AlignCenter, true);
-	AddSText(0, 16, _("Sell items"), UiFlags::ColorWhite | UiFlags::AlignCenter, true);
-	AddSText(0, 18, _("Repair items"), UiFlags::ColorWhite | UiFlags::AlignCenter, true);
-	AddSText(0, 20, _("Leave the shop"), UiFlags::ColorWhite | UiFlags::AlignCenter, true);
-	AddSLine(5);
-	storenumh = 20;
-}
-
-void ScrollSmithBuy(int idx)
+void StoreScroll(int boughtitems)
 {
 	ClearSText(5, 21);
 	stextup = 5;
 
+	int idx = 0;
+	for (; boughtitems != 0; idx++) {
+		if (!storeItem[idx].isEmpty())
+			boughtitems--;
+	}
+
 	for (int l = 5; l < 20; l += 4) {
-		if (!smithitem[idx].isEmpty()) {
-			UiFlags itemColor = smithitem[idx].getTextColorWithStatCheck();
+		if (idx >= storenumh)
+			break;
 
-			if (smithitem[idx]._iMagical != ITEM_QUALITY_NORMAL) {
-				AddSText(20, l, smithitem[idx]._iIName, itemColor, true);
-			} else {
-				AddSText(20, l, smithitem[idx]._iName, itemColor, true);
-			}
+		if (storeItem[idx].isEmpty()) {
+			l -= 4;
+			continue;
+		}
 
-			AddSTextVal(l, smithitem[idx]._iIvalue);
-			PrintStoreItem(&smithitem[idx], l + 1, itemColor);
-			stextdown = l;
-			idx++;
+		UiFlags itemColor = storeItem[idx].getTextColorWithStatCheck();
+
+		if (storeItem[idx]._iMagical != ITEM_QUALITY_NORMAL && storeItem[idx]._iIdentified) {
+			AddSText(20, l, storeItem[idx]._iIName, itemColor, true);
+			AddSTextVal(l, storeItem[idx]._iIvalue);
+		} else {
+			AddSText(20, l, storeItem[idx]._iName, itemColor, true);
+			AddSTextVal(l, storeItem[idx]._ivalue);
 		}
+
+		PrintStoreItem(&storeItem[idx], l + 1, itemColor);
+
+		stextdown = l;
+		idx++;
 	}
 
+	stextsmax = std::max(storenumh - 4, 0);
 	if (stextsel != -1 && !stext[stextsel]._ssel && stextsel != BackButtonLine())
 		stextsel = stextdown;
 }
 
-void StartSmithBuy()
+/**
+ * @brief Purchases an item from a vendor.
+ */
+bool StartBuy()
 {
 	stextsize = true;
 	stextscrl = true;
 	stextsval = 0;
+	stextsmax = 20;
 
-	/* TRANSLATORS: This text is white space sensitive. Check for correct alignment! */
-	strcpy(tempstr, fmt::format(_("I have these items for sale:             Your gold: {:d}"), Players[MyPlayerId]._pGold).c_str());
+	storenumh = 0;
+	for (int i = 0; i < storeItemCount; i++) {
+		if (storeItem[i].isEmpty())
+			continue;
 
-	AddSText(0, 1, tempstr, UiFlags::ColorWhitegold | UiFlags::AlignCenter, false);
-	AddSLine(3);
-	ScrollSmithBuy(stextsval);
-	AddItemListBackButton();
+		WitchBookLevel(storeItem[i]);
+		storeItem[i]._iStatFlag = StoreStatOk(storeItem[i]);
 
-	storenumh = 0;
-	for (int i = 0; !smithitem[i].isEmpty(); i++) {
 		storenumh++;
 	}
+	if (storenumh == 0) {
+		return false;
+	}
+
+	AddSText(0, 1, tempstr, UiFlags::ColorWhitegold | UiFlags::AlignCenter, false);

```

Comment:
We need to assign `tempstr` here.
(for example move `sprintf(tempstr, "I have these items for sale: Your gold: %i", MyPlayer->_pGold);`)
Cause it can be override elsewhere.
And we can reenter `StartBuy` after an item is bought.

Result (see title `8`):
![grafik](https://user-images.githubusercontent.com/25415264/145713730-63226da3-8f9f-4b4d-886e-d630eb82bc30.png)